### DISCUSSION TITLE
Enhance store to return side effects as data (i.e. commands)

### DISCUSSION THREAD

**Discussion by BrainCrumbz** [spierala/mini-rx-store#148] [Ideas]
Hi
we're fiddling with MiniRx (thanks for the library!) aiming to obtain something _à-la_ Elm, i.e. a state management where any reducer can return both a new state and one or more _commands_. Commands should be intended as pure data, representing side-effects, that would then be notified/published/subscribed to/... in order to be actually executed.

On Redux something similar has been done by [redux-loop](https://github.com/redux-loop/redux-loop), where reducer can do that (although commands also contain functions to be invoked, so there's no _effect listener_, but anyway). Redux-loop uses a redux feature called _enhancer_ that allows to enrich the whole store in a way similar to how middleware enriches the dispatch phase. Redux-loop _lifts_ the initial [reducer](https://github.com/redux-loop/redux-loop/blob/326e850df3492e3a429a5029641c7f7540672394/src/install.js#L17) in order to extract the commands returned, and with that creates a new _enhanced_ [store](https://github.com/redux-loop/redux-loop/blob/326e850df3492e3a429a5029641c7f7540672394/src/install.js#L24) to be passed further down.

Would something similar to this be possible through a custom MiniRx extension and/or meta-reducer?

At a first look, that seemed the case. One of the first steps we found when toying around was that there's a need for a new reducer signature, something like:

```ts
type CmdReducer<StateType> = (state: StateType, action: Action) => WithCmd<StateType>;

export interface WithCmd<T> {
  state: T,
  cmds: Cmd[],
}

export interface Cmd {
  // TODO
}
```

but just this first step seems blocking to us, as it would not fit into reducer signature when passed to `configureStore`.  
If we could move past this roadblock, then we could write a reducer like:

```ts
export const userReducer: Reducer<UserState> = function (
  state: UserState = initialState,
  action: Action
): WithCms<UserState> {
  switch (action.type) {
    //...
    case UserActionType.AddUser:
      return {
        state: {
          ...state,
          users: [...state.users, userAction.user],
        },
        cmds: [UserCommands.SaveUsers],
      };
  }
}
```

Then the next step would be a working meta-reducer, starting from something like:

```ts
function commandMetaReducer(reducer: Reducer<AppState>): Reducer<AppState> {
  return (state, action) => {
    const result = reducer(state, action);

    const nextState = result.state;
    const cmds = result.cmds;

    processCommands(cmds);

    return nextState;
  }
}
```

Anyway, that would still require a change in how slice reducers are combined together in order to create the overall app reducer.
When executed, all commands returned by slice reducers should be collected to form a single command array, in the same way as all next state slices concur to form the next app state. To this purpose, Redux-loop has custom functions to [combine reducers](https://github.com/redux-loop/redux-loop/blob/326e850df3492e3a429a5029641c7f7540672394/src/combine-reducers.js) and to [reduce reducers](https://github.com/redux-loop/redux-loop/blob/326e850df3492e3a429a5029641c7f7540672394/src/reduce-reducers.js).

Hope this is not too long.
Any suggestions? Thanks

**Comment by spierala** [comment]
Hi @BrainCrumbz ,

The Redux-Loop approach sounds interesting. 
Currently MiniRx does not support that approach, as you pointed out already.
MiniRx uses Actions in Effects to trigger side effects like API calls. With the advantage that you can use RxJS flattening operators to handle race conditions (e.g. switchMap). 

Maybe you could try to still use the current Reducer signature and put the cmds prop on the state object. Not ideal, but it will unblock your first step. 

You could also fork MiniRx and adjust it to your needs. 
One of the advantages of MiniRx is, that there is not much code, and we tried to make that code as clear as possible.
I think that you know already quite good, what needs to be adjusted.

If you have a working prototype, we can still see if some ideas can go into the MiniRx repo.
Maybe it is possible to make MiniRx flexible enough to support the redux-loop approach (maybe with a configurable combineReducers).
That would be cool, at the same time MiniRx should not become much more complex / less lightweight to support a rather edge-case scenario.

Florian

**Comment by BrainCrumbz** [comment]
Thanks for your reply. We understand your points.

You're right, our idea too was to initially put the cmds on the state object. I can imagine that an _extended_ `configureStore` could let you pass a custom combineReducers: this may somehow become an extension point in the future, still keeping compatibility with original library. The new `configureStore` might also allow for the different reducer signature, so again somehow being compatible or an extension of the original library.

But given that store configuration goes right to `StoreCore`, with its internal calls, I don't know, it seems quite daunting to make it an extension of original library. It looks like we would duplicate & modify StoreCore, or something else to _open it up_.

Anyway, if you have hints on which point should be touched in order to provide this reducer-returning-commands feature, it would be great. Thanks again

**Comment by spierala** [comment]
I would suggest to fork MiniRx and create a version which fully supports your use-case, maybe **only** your use-case in the first place.
When you have a working prototype, we can see if it's worth it to make MiniRx more flexible to support redux-loop style. 

Most likely you will have to adjust `combineReducers`. 
Maybe other places, you will find out when creating the prototype :)

When you forked:

npm i
npm run dev

dev will start up the Angular demo app with a watch on the lib as well.

Then enjoy coding! 

Let me know if you have questions about the current MiniRx implementation.

**Comment by BrainCrumbz** [comment]
Question about current [Effects in Store (Redux API)](https://mini-rx.io/docs/effects):

when an Effect needs access to current store state, what is the suggested best practice in Mini-Rx? I guess it should be some kind of [withLatestFrom](https://www.learnrxjs.io/learn-rxjs/operators/combination/withlatestfrom), but where's the getState observable? We could not see any example.

For action observable there's `actions$`, but for state? Thanks

**Comment by spierala** [comment]
Indeed, you have to use `withLatestFrom`

`store.select` returns an Observable


small example:

```typescript

const myFeatureState$ = store.select(storeState => storeState.myFeature);

const loadEffect = actions$.pipe(
  ofType(TodoActionTypes.LoadTodos),
  withLatestFrom(myFeatureState$),
  mergeMap(([_, myFeatureState]) =>
    // do something with myFeatureState
    ajax<Todo[]>('https://jsonplaceholder.typicode.com/todos').pipe(
      map(res => new LoadTodosSuccess(res.response)),
      catchError(err => of(new LoadTodosFail(err)))
    )
  )
);
```

You might want to look into the Selectors section of the docs. There you can read more about selecting state from the store: https://mini-rx.io/docs/selectors

**Comment by BrainCrumbz** [comment]
We had our go at extending mini-rx-store with reducers that returns commands in order to request side-effects.

The branch of interest is [here](https://github.com/BrainCrumbz/mini-rx-store/tree/feat/return-side-effects). Current comparison with mini-rx-store master is [here](https://github.com/spierala/mini-rx-store/compare/master...BrainCrumbz:feat/return-side-effects).

We added a new example application (mini-rx-vanilla-demo), not using any UI framework. This is the reason for the apparent high number of files. You might want to skip that, or maybe only look at its `app-store.ts` and `user-cmd-effects.ts`. Those command-based effects are basically the same as the action-based ones in `user-effects.ts`.

We did some changes in `libs/mini-rx-store/src/index.ts` to avoid some funny webpack warnings. If you are not experiencing those, they can be ignored.

`libs/mini-rx-store/src/lib/store-core.ts` was modified in a temporary way in order to pass another combineReducer. This is very likely the only place that would require some kind of change within mini-rx-store to let this whole functionality work. Everything else should be obtained as a third-party extension.

There are 4 new files under `libs/mini-rx-store/src/lib/extensions/command` containing the whole feature: store extension, meta reducer, combine-reducers and a command processor with the actual work (to publish commands and to subscribe effects). Again, those could be potentially moved outside of mini-rx-store. There should be no reference to any internal mini-rx type.

There's an optional change that could be made in mini-rx-store to ease the job of this functionality: to provide some kind of `isAction` function relying e.g. on a custom Symbol. At the moment we went by elimination: if it's not undefined, and is not a command, and has a string type, then it must be an action. Not ideal, but somehow brittle: it works until the effect returns something else, even by mistake.

What are your thoughts about it?

**Comment by spierala** [comment]
This is great! I created a Draft PR from your branch and added some comments/questions there: https://github.com/spierala/mini-rx-store/pull/155

I can imagine to make combineReducers configurable in MiniRx.
You could create a dedicated library in another repo for the redux-loop extension code.

**Comment by BrainCrumbz** [comment]
Thanks. We added some comment there. Just be on the same page: we _don't_ expect you to add the extension code to this repository. Only the needed changes to mini-rx-store, like the ability to customize combineReducers.

**Reply by spierala** [reply]
Yes, we are on the same page :)

Feel free to create a PR with the configurable combineReducers fn.

That would be a great addition to make MiniRx even more flexible.

**Comment by spierala** [comment]
@BrainCrumbz I mentioned your contribution on Mastodon: https://mas.to/@spierala/109397931003153516

**Reply by BrainCrumbz** [reply]
Thank you @spierala !

**Comment by spierala** [comment]
@BrainCrumbz 

hey @BrainCrumbz, are you still using the custom combineReducers with MiniRx?

If you are not using it, then I would maybe remove that feature in mini-rx-store@6

Greetz

**Reply by BrainCrumbz** [reply]
Hello there. We're not actively using that feature, so I guess you can proceed.

Thanks for asking!

**Reply by spierala** [reply]
OK thanks,

if you ever need it again, we can add it.
